smolyay/samples.py

```diff
@@ -508,3 +508,620 @@ def _create(self):
+        return scipy.stats.qmc.scale(p_gen, lower_bounds, upper_bounds)
+
+
+class HaltonRandomPointSet(QMCRandomPointSet):
+    """Generates a grid using Halton Sequences
+
+    Parameters
+    ----------
+    domain: list
+        domain of the random points
+
+    number_points: int
+        number of points to generate
+
+    seed: {int, numpy.random.Generator}
+        seed for generating the random points
+
+    scramble : bool
+        Default True. Applies centering to Latin Hypercube points, Owen
+        scrambling to Halton points, and LMS+shift scrambling to Sobol points.
+
+    optimization : {None, "random-cd", "lloyd"}
+        Default None. If "random-cd" the coordinates of points are adjusted to
+        lower the centered discrepancy. If "lloyd", adjust points using a
+        Lloyd-Max algorithm to encourage even spacing.
+
+    bits : int
+        Default 30. Sets the max number of points that can be generated,
+        which is 2**bits.
+    """
+
+    def _get_random_points(self):
+        lower_bounds = [bound[0] for bound in self.domain]
+        upper_bounds = [bound[1] for bound in self.domain]
+        p_gen = scipy.stats.qmc.Halton(
+            self.num_dimensions,
+            scramble=self.scramble,
+            optimization=self.optimization,
+            seed=self.seed,
+        ).random(n=self.number_points)
+        return scipy.stats.qmc.scale(p_gen, lower_bounds, upper_bounds)
+
+
+class SobolRandomPointSet(QMCRandomPointSet):
+    """Generates a grid using Sobol Sequence
+
+    Parameters
+    ----------
+    domain: list
+        domain of the random points
+
+    number_points: int
+        number of points to generate. Must be a power of 2.
+
+    seed: {int, numpy.random.Generator}
+        seed for generating the random points
+
+    scramble : bool
+        Default True. Applies centering to Latin Hypercube points, Owen
+        scrambling to Halton points, and LMS+shift scrambling to Sobol points.
+
+    optimization : {None, "random-cd", "lloyd"}
+        Default None. If "random-cd" the coordinates of points are adjusted to
+        lower the centered discrepancy. If "lloyd", adjust points using a
+        Lloyd-Max algorithm to encourage even spacing.
+
+    Raises
+    ------
+    ValueError
+        number of points must be a power of two
+    ValueError 
+        number of points must be less than 2**bits
+    """
+
+    def __init__(
+        self, domain, number_points, seed, scramble=True, optimization=None, bits=30
+    ):
+        self._bits = 64  # max value of bits
+        super().__init__(domain, number_points, seed, scramble, optimization)
+        self.bits = bits
+
+    @property
+    def bits(self):
+        """{1, 2}: Whether to create an orthogonal array of points"""
+        return self._bits
+
+    @bits.setter
+    def bits(self, value):
+        bits = int(value)
+        if bits > 64:
+            raise ValueError("bits max value is 64.")
```

Comment:
You need to check for a positive power of 2, right? This is the bitwise math to do that.
```suggestion
        if bits > 64:
            raise ValueError("bits max value is 64.")
        elif bits < 1:
            raise ValueError("bits must be at least 1")
        elif bits & (bits-1) != 0:
            raise ValueError("bits must be a power of 2")
```

Reply:
`bits` doesn't have to be a power of 2. It's default value is 30, after all. `bits` is related to the maximum number of points before the Sobol sequence repeats itself which the QMCEngine doesn't allow, and that max number is `2**bits`.

Reply:
Oh, I misread the errors above that about bits and points. You can remove this check on power of 2 here, but this is how you should do it for `num_points` below. It is exact because it is bitwise integer math.